c++/src/kj/string.h

```diff
@@ -514,27 +642,41 @@ StringPtr strPreallocated(ArrayPtr<char> buffer, Params&&... params) {
   //     char buffer[256];
   //     StringPtr text = strPreallocated(buffer, params...);
   //
-  // This is useful for optimization. It can also potentially be used safely in async signal
-  // handlers. HOWEVER, to use in an async signal handler, all of the stringifiers for the inputs
-  // must also be signal-safe. KJ guarantees signal safety when stringifying any built-in integer
-  // type (but NOT floating-points), basic char/byte sequences (ArrayPtr<byte>, String, etc.), as
-  // well as Array<T> as long as T can also be stringified safely. To safely stringify a delimited
-  // array, you must use kj::delimited(arr, delim) rather than the deprecated
-  // kj::strArray(arr, delim).
+  // This is useful for optimization. For use in a signal handler use strSignalSafe() instead.
 
   char* end = _::fillLimited(buffer.begin(), buffer.end() - 1,
       toCharSequence(kj::fwd<Params>(params))...);
   *end = '\0';
   return StringPtr(buffer.begin(), end);
 }
 
+template <typename... Params>
+StringPtr KJ_ALWAYS_INLINE(strSignalSafe(ArrayPtr<char> buffer, Params&&... params)) {
```

Comment:
`strSignalSafe` just has some static asserts & otherwise delegates to `strPreallocated`, hence the `KJ_ALWAYS_INLINE`. I suspect `strPreallocated` should be annotated never inline. I'd need to rewatch https://channel9.msdn.com/Events/CPP/C-PP-Con-2014/0004-Lightning-Talks-Pimp-my-Log before I make that change though - in the past I've applied this logging template optimization to a surprisingly significant reduction in binary size (like 10%). I'm curious to see what this amounts to for KJ when I get around to it.

Reply:
Generally the advice I've heard around always_inline is that you should never apply it just because you think, intuitively, that the function should always be inlined. Instead, you should trust the compiler, unless you have actually observed that the compiler does not do the best thing.